### PR TITLE
full support non rust naming conventions of generated code

### PR DESCRIPTION
**Pull request by framlog**
+ related issue: #562, #565

No difference with current version if c++ namespace is enabled. But it generates a root mod to wrap all codes otherwise.

### CODE REVIEW

File: src/codegen/mod.rs [hunk 1]
```diff
@@ -360,6 +364,12 @@ impl CodeGenerator for Module {
                 }
             }
 
+            if item.id() == ctx.sentinel_module() {
```
Comment (emilio) [inline src/codegen/mod.rs]: The whole `sentinel_module` thing really feels like a hack. Is it just for this? If so, if you always want this at the top of the file, let's just move its handling above.

File: src/codegen/mod.rs [hunk 2]
```diff
@@ -370,13 +380,11 @@ impl CodeGenerator for Module {
...
                 }
-                if result.saw_objc {
-                    utils::prepend_objc_header(ctx, &mut *result);
-                }
             }
         };
 
-        if !ctx.options().enable_cxx_namespaces ||
+        let name = item.canonical_name(ctx);
+        if name == "<sentinel>" || (name != "root" && !ctx.options().enable_cxx_namespaces) ||
```
Comment (emilio) [inline src/codegen/mod.rs]: Just check the `item_id`?

File: src/ir/function.rs [hunk 3]
```diff
@@ -131,7 +131,8 @@ pub fn cursor_mangling(ctx: &BindgenContext,
     }
 
     // Try to undo backend linkage munging (prepended _, generally)
-    if cfg!(target_os = "macos") {
+    if get_abi(cursor.cur_type().call_conv()).map_or(true, |abi| abi != abi::Abi::Stdcall) &&
```
Comment (emilio) [inline src/ir/function.rs]: This only seems to be working around a test expectation difference on OSX while running with `-target msvc`, but this is not a proper fix.

The difference happen because this `cfg!` check is on the target that compiled bindgen, not on the target that clang is parsing files as. That's a totally unrelated fix, and you should just skip adding that file to the diff and revert this change.

File: src/lib.rs [hunk 4]
```diff
@@ -792,6 +792,10 @@ impl<'ctx> Bindings<'ctx> {
             try!(writer.write("\n".as_bytes()));
         }
 
+        if !self.context.options().enable_cxx_namespaces && self.module.items.len() > 0 {
```
Comment (emilio) [inline src/lib.rs]: nit: `!self.module.items.is_empty()`.

File: src/codegen/mod.rs (hunk 1 above)
Comment (framlog) [inline src/codegen/mod.rs]: Yeah, the `sentinel_module` is only used for this. It's designed to be a virtual top mod for the whole code.  But, anyway, I don't understand what's your point of "move its handling above".

File: src/ir/function.rs (hunk 3 above)
Comment (framlog) [inline src/ir/function.rs]: I don't think its a fix either. In fact, I don't fully understand the purport of this if statement since I    got the proper names on macos when the calling convention is `stdcall` or `cdecl` or whatever... I'll remove this line.

**Comment by framlog** [comment]
Thanks for reviewing. I think it's just a small hack except one should pay little attention to handle root mod when cxx namespace is enabled...

**Comment by fitzgen** [comment]
@framlog hi :)

Are you still planning on rebasing this and addressing @emilio's feedback?

**Comment by framlog** [comment]
@fitzgen sry, I thought I've done....

File: src/lib.rs [hunk 5]
```diff
@@ -826,6 +826,10 @@ impl<'ctx> Bindings<'ctx> {
             try!(writer.write("\n".as_bytes()));
         }
 
+        if !self.context.options().enable_cxx_namespaces && !self.module.items.is_empty() {
+            try!(writer.write("pub use self::root::*;\n\n".as_bytes()));
```
Comment (emilio) [inline src/lib.rs]: Let's use `write_str` here.

File: src/codegen/mod.rs [hunk 6]
```diff
@@ -371,14 +381,12 @@ impl CodeGenerator for Module {
...
-                if result.saw_objc {
-                    utils::prepend_objc_header(ctx, &mut *result);
-                }
             }
         };
 
-        if !ctx.options().enable_cxx_namespaces ||
-           (self.is_inline() && !ctx.options().conservative_inline_namespaces) {
+        let name = item.canonical_name(ctx);
+        if name == "<sentinel>" || (name != "root" && !ctx.options().enable_cxx_namespaces) ||
```
Comment (emilio) [inline src/codegen/mod.rs]: We should check the module id's instead of the names.

File: src/codegen/mod.rs [hunk 7]
```diff
@@ -1258,7 +1265,7 @@ impl CodeGenerator for CompInfo {
                     quote_ty!(ctx.ext_cx(), __BindgenUnionField<$ty>)
                 }
             } else if let Some(item) =
-                field_ty.is_incomplete_array(ctx) {
+            field_ty.is_incomplete_array(ctx) {
```
Comment (emilio) [inline src/codegen/mod.rs]: Please remove the rustfmt changes so this is easier to review.

**Comment by jdm** [comment]
@emilio Will you have time to get back to this?

**Comment by emilio** [comment]
So the thing is I'm still not quite sure this is the right trade off, both in terms of the complexity it introduces into bindgen, and in the generated output.

I still think it's quite easy to just use the `raw_line` API so that each person can adapt bindgen to their needs.

Now we have a tools team, so I guess we could discuss this, wdyt @fitzgen?

**Comment by fitzgen** [comment]
I think we've kind of come to the conclusion that this isn't worth the trouble, and its pretty out of date now.

Going to close it and we can open new issues/PRs for anything more in the future
